lib/charms/mongodb/v0/shards_interface.py

```diff
@@ -103,6 +170,28 @@ def _update_relation_data(self, relation_id: int, data: dict) -> None:
             if relation:
                 relation.data[self.charm.model.app].update(data)
 
+    def _get_shards_from_relations(self, departed_shard_id: Optional[int]):
+        """Returns a list of the shards related to the config-server."""
+        relations = self.model.relations[self.relation_name]
+        return set(
+            [
+                self._get_shard_name_from_relation(relation)
+                for relation in relations
+                if relation.id != departed_shard_id
+            ]
+        )
+
+    def _get_shard_hosts(self, shard_name) -> str:
+        """Retrieves the hosts for a specified shard."""
+        relations = self.model.relations[self.relation_name]
+        for relation in relations:
+            if self._get_shard_name_from_relation(relation) == shard_name:
+                return json.loads(relation.data[relation.app].get(HOSTS_KEY, "[]"))
```

Comment:
Can we get a list of hosts from the relation itself (e.g. `relation.units`)? without passing data via relation

Reply:
I would hope so, but `relation.units` doesn't seem to give us the information about the units. I asked [the ops team](https://chat.charmhub.io/charmhub/pl/m7jmq3wasfg6dj8m945mqseosa)

Reply:
From what I can tell from looking at [`ops.model.Unit`](https://ops.readthedocs.io/en/latest/_modules/ops/model.html#Unit) there is no straightforward way to get the IP address. I assume that what we have here is the accepted way of doing things. (After-all this keeps with the pattern we use for client relations in Data Platform Interface)

Reply:
yes